src/System.IdentityModel.Tokens/AsymmetricSignatureProvider.cs

```diff
@@ -50,62 +90,109 @@ public class AsymmetricSignatureProvider : SignatureProvider
+#endif
+            if (!algorithmResolved)
+                throw new ArgumentOutOfRangeException(string.Format(CultureInfo.InvariantCulture, LogMessages.IDX10641, key.ToString()));
+        }
+
+        /// <summary>
+        /// Gets the mapping from algorithm to the minimum <see cref="AsymmetricSecurityKey"/>.KeySize for creating signatures.
+        /// </summary>
+        public IReadOnlyDictionary<string, int> MinimumAsymmetricKeySizeInBitsForSigningMap
+        {
+            get
+            {
+                return minimumAsymmetricKeySizeInBitsForSigningMap;
             }
+        }
 
-            if (key.KeySize < SignatureProviderFactory.MinimumAsymmetricKeySizeInBitsForVerifying)
+        /// <summary>
+        /// Gets the mapping from algorithm to the minimum <see cref="AsymmetricSecurityKey"/>.KeySize for verifying signatures.
+        /// </summary>
+        public IReadOnlyDictionary<string, int> MinimumAsymmetricKeySizeInBitsForVerifyingMap
+        {
+            get
             {
-                throw new ArgumentOutOfRangeException("key.KeySize", key.KeySize, string.Format(CultureInfo.InvariantCulture, LogMessages.IDX10630, key.GetType(), SignatureProviderFactory.MinimumAsymmetricKeySizeInBitsForVerifying));
+                return minimumAsymmetricKeySizeInBitsForVerifyingMap;
             }
+        }
+
 
 #if DNXCORE50
+        protected virtual HashAlgorithmName GetHashAlgorithmName(string algorithm)
+        {
+            if (string.IsNullOrWhiteSpace(algorithm))
+                LogHelper.Throw(string.Format(CultureInfo.InvariantCulture, LogMessages.IDX10000, "GetHashAlgorithmName.algorithm"), typeof(ArgumentNullException), EventLevel.Verbose);
+
+            switch (algorithm)
+            {
+                case SecurityAlgorithms.SHA256:
+                case SecurityAlgorithms.ECDSA_SHA256:
+                case SecurityAlgorithms.RSA_SHA256:
+                case SecurityAlgorithms.RsaSha256Signature:
+                    return HashAlgorithmName.SHA256;
+
+                case SecurityAlgorithms.SHA384:
+                case SecurityAlgorithms.ECDSA_SHA384:
+                case SecurityAlgorithms.RSA_SHA384:
+                case SecurityAlgorithms.RsaSha384Signature:
+                    return HashAlgorithmName.SHA384;
+
+                case SecurityAlgorithms.SHA512:
+                case SecurityAlgorithms.ECDSA_SHA512:
+                case SecurityAlgorithms.RSA_SHA512:
+                case SecurityAlgorithms.RsaSha512Signature:
+                    return HashAlgorithmName.SHA512;
+
+                default:
+                    throw new ArgumentOutOfRangeException(string.Format(CultureInfo.InvariantCulture, LogMessages.IDX10640, algorithm));
+            }
+        }
+
+        private bool ResolveDotNetCoreAsymmetricAlgorithm(AsymmetricSecurityKey key, string algorithm, bool willCreateSignatures)
+        {
             hash = GetHashAlgorithmName(algorithm);
```

Comment:
We don't need to create the hash if clr returns RSACryptoServiceProvider because Sign and VerifyData take the hash.Name. Is there a way to just use the name?